### DISCUSSION TITLE
Switch to Build System

### DISCUSSION THREAD

**Discussion by BttrDrgn** [mholtkamp/octave#26] [Ideas]
I personally believe it would be a good investment to switch to a build system that can handle automatically setting up the required small tweaks required when using the solution, like having to set the debug runtime directory to $(SolutionDir), pre-compile the shaders properly per os, and other tweaks you can do to generate the solution and other ide project files rather than having just the solution for Windows, also more easily cross-platform and ide/tool compilable.

I've only really used premake (https://premake.github.io/) and enjoy using it but CMake (https://cmake.org/) may be a better choice since it is more robust.

I can provide the changes and submit a pr, but I wanted to discuss first before diving in.

**Comment by mholtkamp** [comment]
Yes switching to a proper build system would be ideal. I don't have much experience with build systems though, and I'm unsure how much work it would be to support the homebrew / Android toolchains. But yes it would be so much better than maintaining all of the separate Makefiles + VS project. And also I don't have a good way of building "Release" and "Debug" configurations for Linux, which is a pain.

I'm open to both Premake and CMake. Premake sounds like it might be a bit easier? People seem to really like it :) But yes CMake is more common and is already being used for the Android build.

**Comment by BttrDrgn** [comment]
Premake can be used to run external commands too and you can always extend the build system by writing additional platforms into it. For example, Blizzard wrote an extension for Xbox One/Series (https://github.com/Blizzard/premake-consoles), it's pretty modular.

I'll fire up a fork and see what I can do with premake, if it comes to it ill learn cmake though!

**Reply by mholtkamp** [reply]
Awesome! Yeah that sounds good. Thanks for bringing up the discussion and looking into it.

**Reply by mardy** [reply]
Just found this thread. I know cmake relatively well, and although it's not my favourite build system (I do prefer [meson](https://mesonbuild.com/)), I could give it a try, if you are still interested.

**Reply by mardy** [reply]
I'm having some trouble understanding what External/Assimp is :-) It does have a CMakeLists.txt file, which does not match the project structure (for example, it expects to find WafeFront files in `Obj/`, whereas they are in `WavefrontObj/`), so it seems that somehow the Assimp files went out of sync. So I checked the upstream repository, but even looking at old releases (~2019), the project structure is very different from what we have here.

@mholtkamp, do you remember what version of Assimp you have imported? Do you think we could switch to a model where external dependencies are used as-is (maybe imported as git submodules, so it will be easier to update them), or do you have a lot of local changes that need to be preserved? At a first look, my impression is that (at least as far as Assimp is concerned) the local changes are compilation fixes that are most likely already present in new Assimp releases -- but I might have missed something.

**Reply by mholtkamp** [reply]
Hey Mardy! Yeah sorry about Assimp, it is a bit of a mess. I originally had to make changes to the source code to get it to compile, but I think it's because I didnt' configure it correctly with CMake when I first started. I believe I used version 5.0. I think we could try upgrading to the latest version of assimp.
